ehrql/assurance.py

```diff
@@ -69,11 +56,33 @@ def validate(variable_definitions, test_data):
     }
 
     # Validate results of query
-    return {
+    test_validation_errors = {
         patient_id: validation_result
         for patient_id, patient in test_data.items()
         if (validation_result := validate_patient(patient_id, patient, query_results))
     }
+    return {
+        "constraint_validation_errors": constraint_validation_errors,
+        "test_validation_errors": test_validation_errors,
+    }
+
+
+def validate_constraints(records, table):
+    unexpected_test_values = []
+    for record in records:
+        for column, value in record.items():
+            constraints = table.schema.schema[column].constraints
+            for constraint in constraints:
+                is_valid = constraint.validate(value)
+                if not is_valid:
+                    unexpected_test_values.append(
+                        {
+                            "column": column,
+                            "constraint": f"{constraint}",
+                            "value": f"{value}",
```

Comment:
I should probably use `str()` here instead of f string here for simplicity?

Reply:
Yeah, maybe. I think the f-string version is pretty clear though – I wouldn't have picked it out as a thing to change myself.

Reply:
thx, I'll keep it as it is then!